Feature/multi monitor

AdamWagner opened the pull request:
I'm pleasantly surprised that multi-monitor support was possible without manually keeping track of screens (and maybe spaces)! 

@alin23 would be awesome if you could try this out sometime. This this branch is going to be my daily driver from now on, so I'll be testing as well ;-)

alin23 reviewed (approved):
This works for me on both monitors, good job on this 👏 

There are a few small issues to fix, I've commented on the line for the missing `getFocusedScreen`.
There's also an issue with the `require` line in init.lua. The `stackline` variable should be global so `local` should be removed:

```sh
stackline = require "stackline.stackline.stackline"
```

Another thing that would be great to have is configurable border radius (if it isn't already). `width / 3` is a bit too rounded for the Big Sur icons

I also found some issues in StackConfig:

Here, `stackConfig` should be `stackline.config`
```lua
        if key == "toggle_icons" then
            stackConfig:toggle('showIcons') -- global var
        end
```

And here, `Sm` should be `stackline.manager`
```lua
function StackConfig:registerWatchers()
    local key = 'showIcons'
    local identifier = self:makePath(key .. '-handler')
    local settingPath = self:makePath(key)
    self.store.watchKey(identifier, settingPath, function(_val)
        Sm:toggleIcons()
    end)
    return self
end
```

johnallen3d commented:
I feel like the problem child here... trying this out I'm not seeing indicators but there are a fair number of errors in my Hammerspoon console. Perhaps these errors are related to what @alin23 has already reported but that's not clear to me (no Lua or Hammerspoon API experience).

```txt
2020-08-26 22:45:16: Welcome to the Hammerspoon Console!
You can run any Lua code in here.

2020-08-26 22:45:16: -- Lazy extension loading enabled
2020-08-26 22:45:16: -- Loading ~/.hammerspoon/init.lua
2020-08-26 22:45:16: -- Loading extension: settings
2020-08-26 22:45:16: org.hammerspoon.Hammerspoon
2020-08-26 22:45:16: -- Loading extension: fnutils
2020-08-26 22:45:16: -- Loading extension: ipc
2020-08-26 22:45:16: -- Loading extension: window
2020-08-26 22:45:16: -- Loading extension: uielement
2020-08-26 22:45:16: -- Loading extensions: window.filter
2020-08-26 22:45:16: -- Loading extension: timer
2020-08-26 22:45:16: -- Loading extension: spaces
2020-08-26 22:45:16: -- Done.
2020-08-26 22:45:16: 



2020-08-26 22:45:16: ========================================
2020-08-26 22:45:16: STARTING STACKLINE	==========
2020-08-26 22:45:16: ========================================
2020-08-26 22:45:16: ********
2020-08-26 22:45:16: 22:45:16 ERROR:   LuaSkin: hs.timer callback error: .../johna/.hammerspoon/stackline/stackline/stackmanager.lua:12: attempt to index a nil value (global 'stackline')
stack traceback:
	.../johna/.hammerspoon/stackline/stackline/stackmanager.lua:12: in function 'stackline.stackline.stackmanager.new'
	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:24: in function 'stackline.stackline.stackline.start'
	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:85: in upvalue 'actionFn'
	...poon.app/Contents/Resources/extensions/hs/timer/init.lua:147: in function <...poon.app/Contents/Resources/extensions/hs/timer/init.lua:145>
2020-08-26 22:45:16: ********
2020-08-26 22:45:21: ********
2020-08-26 22:45:21: 22:45:21 ERROR:   LuaSkin: hs.application.watcher callback: /Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: attempt to index a nil value (field 'manager')
stack traceback:
	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: in function 'stackline.stackline.stackline.redrawWinIndicator'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:944: in upvalue 'emit'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:970: in method 'filterEmitEvent'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1004: in method 'emitEvent'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1096: in method 'unfocused'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1285: in method 'deactivated'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1270: in function <...n.app/Contents/Resources/extensions/hs/window/filter.lua:1267>
	(...tail calls...)
2020-08-26 22:45:21: ********
2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: Hammerspoon (10265) is not registered yet
2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP Web Content (no watcher pid)
2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP WebExtensions (no watcher pid)
2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app Ferdi Helper (Renderer) (no watcher pid)
2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP Privileged Content (no watcher pid)
2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP RDD Process (no watcher pid)
2020-08-26 22:45:22: ********
2020-08-26 22:45:22: 22:45:22 ERROR:   LuaSkin: hs.timer callback error: /Users/johna/.hammerspoon/stackline/stackline/stackline.lua:31: attempt to index a nil value (field 'manager')
stack traceback:
	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:31: in function </Users/johna/.hammerspoon/stackline/stackline/stackline.lua:28>
2020-08-26 22:45:22: ********
2020-08-26 22:45:29: ********
2020-08-26 22:45:29: 22:45:29 ERROR:   LuaSkin: hs.application.watcher callback: /Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: attempt to index a nil value (field 'manager')
stack traceback:
	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: in function 'stackline.stackline.stackline.redrawWinIndicator'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:944: in upvalue 'emit'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:970: in method 'filterEmitEvent'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1004: in method 'emitEvent'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1088: in method 'focused'
	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1280: in function <...n.app/Contents/Resources/extensions/hs/window/filter.lua:1267>
	(...tail calls...)
2020-08-26 22:45:29: ********
```

alin23 commented:
> I feel like the problem child here... trying this out I'm not seeing indicators but there are a fair number of errors in my Hammerspoon console. Perhaps these errors are related to what @alin23 has already reported but that's not clear to me (no Lua or Hammerspoon API experience).
> 
> ```
> 2020-08-26 22:45:16: Welcome to the Hammerspoon Console!
> You can run any Lua code in here.
> 
> 2020-08-26 22:45:16: -- Lazy extension loading enabled
> 2020-08-26 22:45:16: -- Loading ~/.hammerspoon/init.lua
> 2020-08-26 22:45:16: -- Loading extension: settings
> 2020-08-26 22:45:16: org.hammerspoon.Hammerspoon
> 2020-08-26 22:45:16: -- Loading extension: fnutils
> 2020-08-26 22:45:16: -- Loading extension: ipc
> 2020-08-26 22:45:16: -- Loading extension: window
> 2020-08-26 22:45:16: -- Loading extension: uielement
> 2020-08-26 22:45:16: -- Loading extensions: window.filter
> 2020-08-26 22:45:16: -- Loading extension: timer
> 2020-08-26 22:45:16: -- Loading extension: spaces
> 2020-08-26 22:45:16: -- Done.
> 2020-08-26 22:45:16: 
> 
> 
> 
> 2020-08-26 22:45:16: ========================================
> 2020-08-26 22:45:16: STARTING STACKLINE	==========
> 2020-08-26 22:45:16: ========================================
> 2020-08-26 22:45:16: ********
> 2020-08-26 22:45:16: 22:45:16 ERROR:   LuaSkin: hs.timer callback error: .../johna/.hammerspoon/stackline/stackline/stackmanager.lua:12: attempt to index a nil value (global 'stackline')
> stack traceback:
> 	.../johna/.hammerspoon/stackline/stackline/stackmanager.lua:12: in function 'stackline.stackline.stackmanager.new'
> 	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:24: in function 'stackline.stackline.stackline.start'
> 	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:85: in upvalue 'actionFn'
> 	...poon.app/Contents/Resources/extensions/hs/timer/init.lua:147: in function <...poon.app/Contents/Resources/extensions/hs/timer/init.lua:145>
> 2020-08-26 22:45:16: ********
> 2020-08-26 22:45:21: ********
> 2020-08-26 22:45:21: 22:45:21 ERROR:   LuaSkin: hs.application.watcher callback: /Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: attempt to index a nil value (field 'manager')
> stack traceback:
> 	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: in function 'stackline.stackline.stackline.redrawWinIndicator'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:944: in upvalue 'emit'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:970: in method 'filterEmitEvent'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1004: in method 'emitEvent'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1096: in method 'unfocused'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1285: in method 'deactivated'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1270: in function <...n.app/Contents/Resources/extensions/hs/window/filter.lua:1267>
> 	(...tail calls...)
> 2020-08-26 22:45:21: ********
> 2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: Hammerspoon (10265) is not registered yet
> 2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP Web Content (no watcher pid)
> 2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP WebExtensions (no watcher pid)
> 2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app Ferdi Helper (Renderer) (no watcher pid)
> 2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP Privileged Content (no watcher pid)
> 2020-08-26 22:45:21: 22:45:21 ** Warning:   wfilter: No accessibility access to app FirefoxCP RDD Process (no watcher pid)
> 2020-08-26 22:45:22: ********
> 2020-08-26 22:45:22: 22:45:22 ERROR:   LuaSkin: hs.timer callback error: /Users/johna/.hammerspoon/stackline/stackline/stackline.lua:31: attempt to index a nil value (field 'manager')
> stack traceback:
> 	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:31: in function </Users/johna/.hammerspoon/stackline/stackline/stackline.lua:28>
> 2020-08-26 22:45:22: ********
> 2020-08-26 22:45:29: ********
> 2020-08-26 22:45:29: 22:45:29 ERROR:   LuaSkin: hs.application.watcher callback: /Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: attempt to index a nil value (field 'manager')
> stack traceback:
> 	/Users/johna/.hammerspoon/stackline/stackline/stackline.lua:38: in function 'stackline.stackline.stackline.redrawWinIndicator'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:944: in upvalue 'emit'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:970: in method 'filterEmitEvent'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1004: in method 'emitEvent'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1088: in method 'focused'
> 	...n.app/Contents/Resources/extensions/hs/window/filter.lua:1280: in function <...n.app/Contents/Resources/extensions/hs/window/filter.lua:1267>
> 	(...tail calls...)
> 2020-08-26 22:45:29: ********
> ```

This is related to the `local stackline = require "stackline.stackline.stackline"` line in your init.lua.

It should be without the local keyword: `stackline = require "stackline.stackline.stackline"` so that `stackline` is available globally

johnallen3d commented:
@alin23 - thanks! 🤦 might be time for me to study Lua a little.

On my external display! So much more responsive! 
<img width="306" alt="Hammerspoon Console 2020-08-27 08-06-24" src="https://user-images.githubusercontent.com/184915/91440263-5762dc80-e83c-11ea-9451-c9abfcdd6b10.png">

AdamWagner commented:
Awesome, thanks @johnallen3d !

AdamWagner commented on stackline/query.lua:
```diff
@@ -100,25 +94,33 @@ function shouldRestack(new) -- {{{
         return true
     end
+
+    local lastFocusedScreen = Query:getFocusedScreen()
```
That was a reference left over from an old iteration. Thanks for pointing this out, @alin23 !

AdamWagner commented:
@alin23 Thanks for the notes.

- Updated the install instructions in the readme & wiki to change the `local` stackline declaration to global: https://github.com/AdamWagner/stackline/commit/11c41273f9f9a3a5ff2b780172d6c510e4bf7393
- Fixed `Sm:toggleIcons` and `stackConfig` issues: https://github.com/AdamWagner/stackline/commit/76ef09ace3930de5fd06f8949b25e8463e27868a

Re: your note about border-radius configurability – configurability has been on my mind as well. I started an issue here and added your comment: 

> https://github.com/AdamWagner/stackline/issues/33
